azure_monitor_metrics_receiver.go

```diff
@@ -0,0 +1,128 @@
+	ResourceID   string
+	Metrics      []string
+	Aggregations []string
+}
+
+type ResourceGroupTarget struct {
+	resourceGroup string
+	resources     []*Resource
+}
+
+type Resource struct {
+	resourceType string
+	metrics      []string
+	aggregations []string
+}
+
+type AzureClients struct {
+	Ctx						context.Context
+	ResourcesClient         ResourcesClient
+	MetricDefinitionsClient MetricDefinitionsClient
+	MetricsClient           MetricsClient
+}
+
+type Metric struct {
+	Name string
+	Fields map[string]interface{}
+	Tags map[string]string
+}
+
+type azureResourcesClient struct {
+	client *armresources.Client
+}
+
+type ResourcesClient interface {
+	List(context.Context, *armresources.ClientListOptions) ([]*armresources.ClientListResponse, error)
+	ListByResourceGroup(context.Context, string, *armresources.ClientListByResourceGroupOptions) ([]*armresources.ClientListByResourceGroupResponse, error)
+}
+
+type MetricDefinitionsClient interface {
+	List(context.Context, string, *armmonitor.MetricDefinitionsClientListOptions) (armmonitor.MetricDefinitionsClientListResponse, error)
+}
+
+type MetricsClient interface {
+	List(context.Context, string, *armmonitor.MetricsClientListOptions) (armmonitor.MetricsClientListResponse, error)
+}
+
+func NewAzureMonitorMetricsReceiver(subscriptionID string, clientID string, clientSecret string, tenantID string, targets *Targets, azureClients *AzureClients) (*AzureMonitorMetricsReceiver, error) {
+	azureMonitorMetricsReceiver := &AzureMonitorMetricsReceiver{
+		Targets: targets,
+		AzureClients: azureClients,
+		subscriptionID: subscriptionID,
+		clientID: clientID,
+		clientSecret: clientSecret,
+		tenantID: tenantID,
+	}
+
+	if err := azureMonitorMetricsReceiver.checkValidation(); err != nil {
+		return nil, fmt.Errorf("got validation error: %v", err)
+	}
+
+	azureMonitorMetricsReceiver.addPrefixToResourceTargetsResourceID()
+	return azureMonitorMetricsReceiver, nil
+}
+
+func NewTargets(resourceTargets []*ResourceTarget, resourceGroupTargets []*ResourceGroupTarget, subscriptionTargets []*Resource) *Targets {
+	return &Targets{
+		ResourceTargets: resourceTargets,
+		resourceGroupTargets: resourceGroupTargets,
+		subscriptionTargets: subscriptionTargets,
+	}
+}
+
+func NewResourceTarget(resourceID string, metrics []string, aggregations []string) *ResourceTarget {
+	return &ResourceTarget{
+		ResourceID:   resourceID,
+		Metrics:      metrics,
+		Aggregations: aggregations,
+	}
+}
+
+func NewResourceGroupTarget(resourceGroup string, resources []*Resource) *ResourceGroupTarget {
+	return &ResourceGroupTarget{
+		resourceGroup: resourceGroup,
+		resources: resources,
+	}
+}
+
+func NewResource(resourceType string, metrics []string, aggregations []string) *Resource {
+	return &Resource{
+		resourceType: resourceType,
+		metrics: metrics,
+		aggregations: aggregations,
+	}
+}
+
+func newAzureResourcesClient(subscriptionID string, credential *azidentity.ClientSecretCredential) *azureResourcesClient {
```

Comment:
Why is this function private? Why does this differ from `NewResource` or `NewResourceTarget` in terms of visibility

Reply:
I use it for polymorphism by using the interface type I created. basically it is for constructor dependency injection, so I can do mocking for unit tests.